client/express-checkout/index.js

```diff
@@ -185,41 +189,70 @@ jQuery( ( $ ) => {
 
 			const eceButton = wcpayECE.createButton( elements, {
 				buttonType: {
-					googlePay: wcpayExpressCheckoutParams.button.type,
-					applePay: wcpayExpressCheckoutParams.button.type,
+					googlePay: getExpressCheckoutData( 'button' ).type,
+					applePay: getExpressCheckoutData( 'button' ).type,
 				},
 			} );
 
 			wcpayECE.showButton( eceButton );
 
-			wcpayECE.attachButtonEventListeners( eceButton );
-
 			eceButton.on( 'click', function ( event ) {
 				const clickOptions = {
-					business: {
-						name: 'Mikes Bikes',
-					},
-					lineItems: [
-						{ name: 'Bike', amount: 200 },
-						{ name: 'Helmet', amount: 300 },
-					],
-					shippingAddressRequired: true,
-					shippingRates: [
-						{
-							id: '1',
-							amount: 500,
-							displayName: 'Standard Shipping',
-						},
-						{
-							id: '2',
-							amount: 1000,
-							displayName: 'Expedited Shipping',
-						},
-					],
+					lineItems: options.displayItems.map( ( i ) => ( {
+						...i,
+						name: i.label,
+					} ) ),
+					emailRequired: true,
+					shippingAddressRequired: options.requestShipping,
+					// FIXME: This is a total hack, we need some way to get the shipping information before rendering the button.
+					//        Possible to just send an empty address maybe? Just make up a free shipping thing since the shipping
+					//        address change event is sent as soon as the payment sheet pops up?
+					shippingRates: options.displayItems
+						.filter( ( i ) => i.label === 'Shipping' )
+						.map( ( i ) => ( {
+							id: `rate-${ i.label }`,
+							amount: i.amount,
+							displayName: i.label,
+						} ) ),
```

Comment:
I'm at a bit of a loss here, not sure how to get the currently selected address, and not even sure what's supposed to happen during a guest session when there is no address available before clicking the button.

Either way the payment sheet UI immediately triggers the `shippingaddresschange` event as soon as it loads, so this might not be a huge problem? Except maybe on slow networks where it's possible that the initial payment method shows up at a lower price than the actual options?

Another thing to consider is what happens when the response is empty. If there is no shipping method available for the currently loaded address then the GPay button will fail to load because if `shippingAddressRequired` is `true` you **must** initialize with a list of `shippingRates`; there is no way to defer loading them until later.

I'm leaning towards:

1. Use the current method to fall back on the "Shipping" display item.
    - This is a little bit brittle in the sense that it could not be present if there are no shipping methods configured for the customer's currently loaded address.
2. If there is no "Shipping" item in the display items: don't try to load the button.

This has the downside of making it so that the button won't load even if the customer is on the checkout page and changes their address to a place that has configured shipping methods. We can work around that by adding listeners to the address inputs on the checkout page but that might be a whole separate area of testing if different themes load those inputs in different ways.

What are your thoughts here @bborman22 in terms of where we should focus the effort here? Maximum compatibility or just the "happy path" of customer has an address loaded that has a corresponding shipping method?

Reply:
cc @rafaelzaleski just in case you faced something similar with the Blocks integration, although I think you always have an address provided by the Blocks API, so you might not have run into this 🤔

Reply:
And I forgot to mention: I think you may get shipping rates from the backend if you send the API an empty address _so long as there is a "rest of world" shipping method configured_. I'm not sure what happens if there is no "rest of world" shipping method configured.

Reply:
I've implemented the approach I'm referring to in 63dabfd6cfcf27b76010f77a9baa8312bf736d55.

Reply:
Just wanted to comment that I'm taking a look here to get a better grasp. Assuming the approach in 63dabfd unblocks progress here, let's move forward with that and create an issue to track any further digging needed. I was trying to take a look and didn't see a 1 to 1 mapping with PRB in this flow, but to summarize we should be targeting feature parity between ECE and PRB if that helps answer the question at all as I'm not sure what happens in those scenarios with PRB yet and will be digging in now to find out.

Reply:
The PRB (current) implementation doesn't require the address upfront to load the buttons, you're instead allowed to wait for the `shippingaddresschange` event to get proper shipping rates, so this is a new problem for ECE.

I think this idea of not rendering the button when we can't find shipping rates is a reasonable middle-ground to "solve" the issue caused by the new API.